src/switch.py

```diff
@@ -0,0 +1,275 @@
+    errorIndication, errorStatus, errorIndex, varBinds = next(
+        getCmd(
+            snmp_engine,
+            snmp['auth_data'],
+            snmp['transport_target'],
+            context_data,
+            ObjectType(ObjectIdentity(OID),)
+        )
+    )
+    if errorIndication:
+        raise SwitchException('Unable to get SNMP value: {}'.
+                              format(errorIndication))
+
+    return convert_snmp_type(varBinds)
+
+
+def get_snmp_table(snmp, OID):
+    ret = {}
+    for errorIndication, errorStatus, errorIndex, varBinds in bulkCmd(
+        snmp_engine,
+        snmp['auth_data'],
+        snmp['transport_target'],
+        context_data,
+        0,
+        25,
+        ObjectType(ObjectIdentity(OID),),
+        lexicographicMode=False,
+        lookupMib=False,
+    ):
+        if errorIndication:
+            raise SwitchException('Unable to get SNMP value: {}'.
+                                  format(errorIndication))
+        index = int(str(varBinds[0][0][-1:]))
+        ret[index] = convert_snmp_type(varBinds)
+
+    return ret
+
+
+def convert_snmp_type(varBinds):
+    val = varBinds[0][1]
+    if type(val) in [
+        proto.rfc1902.Integer,
+        proto.rfc1902.Counter32,
+        proto.rfc1902.Counter64,
+    ]:
+        return int(val)
+    return str(val)
+
+
+def get_switch_model(snmp):
+    model = get_snmp_value(snmp, OIDs['switch_model'])
+
+    if 'PowerConnect' in model:
+        return 'powerconnect'
+    elif 'ProCurve' in model:
+        return 'procurve'
+    elif 'ExtremeXOS' in model:
+        return'extreme'
+    else:
+        print('Unknown switch model', file=sys.stderr)
+    return None
+
+
+def get_monitored_ports(snmp, model):
+    """ Get ports which meet the following conditions:
+        - are configured to be no shutdown
+        - and don't belong to a LAGG
+        - or are a configured LAGG
+
+        SNMP works in misterious ways: it is faster to fetch a whole table
+        and then operate in this program than to fetch just a few entries.
+    """
+
+    ret = {}
+
+    port_names = get_snmp_table(snmp, OIDs['port_name'])
+    laggs = get_laggs(snmp, model)
+
+    # Get only those ports which are up.
+    port_states = {
+        x: y for x, y
+        in get_snmp_table(snmp, OIDs['port_state']).items()
+        if y == 1
+    }
+    for port_idx, port_state in port_states.items():
+        if not laggs or port_idx not in laggs.keys():
+            port_name = standarize_portname(port_names[port_idx])
+            if port_name:
+                ret[port_idx] = port_name
+
+    return ret
+
+
+def get_laggs(snmp, model):
+    """ Get only those LAGGs which have members """
+
+    if model in ['powerconnect', 'procurve']:
+        lagg_oid = LAGG_OIDs[model]
+    else:
+        lagg_oid = None
+
+    if lagg_oid:
+        return {
+            x: y for x, y
+            in get_snmp_table(snmp, lagg_oid).items()
+            if y != 0
+        }
+
+    return None
```

Comment:
Done. Sorry, I was already on process of simplifying this but somehow I missed it.